### DISCUSSION TITLE
Recommend using zigpy-znp over zigpy-cc for CC2530/CC2531 with Z-Stack Home 1.2?

### DISCUSSION THREAD

**Discussion by Hedda** [zigpy/zigpy#618] [ZHA development]
Should [zigpy-znp](https://github.com/zigpy/zigpy-znp) be recommended over [zigpy-cc](https://github.com/zigpy/zigpy-cc) for CC2530/CC2531 coordinator even when using Z-Stack Home 1.2 stack?

Or should the recommendation be to upgrade CC2530/CC2531 coordinators from Z-Stack Home 1.2 stack to Z-Stack 3.0.0 firmware even though the CC2530/CC2531 MCU hardware might be less stable with Z-Stack 3.0.0 stack (due to MCU limited RAM-memory?)?

Regardless maybe zigpy readme-files should be updated and perhaps there could be migration guides for ZHA documentation?

https://www.home-assistant.io/integrations/zha/

ZHA users have reported that devices stopped working with zigpy-cc in Home Assistant 2020.12.1 release and seen that Adminiuga recommended a few Home Assistant ZHA users to which from zigpy-cc to zigpy-znp even when they are presumingly are using CC2531 (or CC2530) based Zigbee coordinator with older Z-Stack Home 1.2 Zigbee stack firmware and not the newer Z-Stack 3.0 stack.

https://github.com/home-assistant/core/issues/44559
https://github.com/home-assistant/core/issues/43268#issuecomment-727628893

A fact to consider is that most people with CC2530 and CC2531 based Zigbee coordinators with Z-Stack Home 1.2 firmware, and other projects like Zigbee2MQTT still recommend Z-Stack Home 1.2 firmware for CC2530 and CC2531, (it should, however, be noted that they recommend Z-Stack 3.0 for CC2538 but then does have a more powerful MCU which features more RAM-memory).

https://github.com/Koenkk/Z-Stack-firmware/tree/master/coordinator
https://www.zigbee2mqtt.io/information/supported_adapters.html

Know that puddly been working on a backup and restore process so wondering if that will work as a migration path?

https://github.com/zigpy/zigpy/issues/557

PS: Can and does both zigpy-znp and zigpy-cc auto-detect and log HW + which version of Z-Stack is running on the hardware?

**Comment by Adminiuga** [comment]
Yes. No.

**Comment by puddly** [comment]
> Should zigpy-znp be recommended over zigpy-cc for CC2530/CC2531 coordinator even when using Z-Stack Home 1.2 stack?

There really isn't anything that interesting with Z-Stack Home 1.2 that requires special handling, I just don't personally use a CC2531 on my home network so all of the testing I do is synthetic and I have to wait for user reports to figure out if something is broken. Z-Stack Home 1.2 doesn't properly support any of the newer features like route re-discovery and manual child aging so support for it is "done" unless firmware changes bring additional functionality, but I doubt this is going to happen.

Maybe in the next few HA releases we can automatically migrate zigpy-cc users over to zigpy-znp, to take advantage of the multi-endpoint, selective joining, and route rediscovery fixes.

> Regardless maybe zigpy readme-files should be updated and perhaps there could be migration guides for ZHA documentation?

I have it indirectly documented in the zigpy-znp README: https://github.com/zigpy/zigpy-znp#home-assistant

> Know that puddly been working on a backup and restore process so wondering if that will work as a migration path?

In theory, yes.

> PS: Can and does both zigpy-znp and zigpy-cc auto-detect and log HW + which version of Z-Stack is running on the hardware?

Z-Stack version detection, approximate hardware detection, and network information logging is in the `dev` branch of zigpy-znp at the moment. I don't have a CC2538 to try it with but I don't believe it's possible to distinguish a CC2531 from a CC2538 using the current command set.

**Reply by balu79** [reply]
Reg. the NVRAM backup, I'm wondering whether that is also available for users of Home Assistant Operating System, so without a Python virtual environment (if I'm correct). I guess not or is there any other way for Home Assistant Operating System users already? Either way, great to see that option and in general the development on zigpy-znp, much appreciated!

**Reply by puddly** [reply]
The NVRAM backup tool is a part of the zigpy-znp Python module so it's already present in HA's Python installation. The problem is that serial communication with the radio is exclusive: you can't run this tool while HA is using the radio. When I last tested this, I found that even briefly shutting down HA will cause the Docker container to restart.

Perhaps it's possible to temporarily rename the `/dev/serial/by-id/...` symlink, restart HA so that it never connects, run one of the tools using the new port path, and then rename the symlink back so that HA can start up normally?

**Reply by MattWestb** [reply]
With bellows with network connected coordinator i editing the `<ZHA_Config_Dir>\.storage\core.config_entries` 
```
            {
                "entry_id": "ca43d86de6df6304c3dc0c305eda4d44",
                "version": 2,
                "domain": "zha",
                "title": "socket://192.168.2.60:8888",
                "data": {
                    "device": {
                        "baudrate": 115200,
                        "path": "socket://192.168.2.60:8888"
                    },
                    "radio_type": "ezsp"
                },
                "options": {},
                "system_options": {
                    "disable_new_entities": false
                },
                "source": "user",
                "connection_class": "local_push",
                "unique_id": null
            }
```
and editing  the key the `"path": ` to somthing that is not working and restarting AH. (only on part of the file i shown)
HA is starting but not ZHA = i can then open one command prompt (CLI) in the container and using the radio lib (for my bellows) without ZHA have locking the comport. And then undoing the changes in the file and restarting HA and its going online without problems.

Perhaps the same trick is working with other radios ??

**Reply by balu79** [reply]
Thanks! Understanding that I can't run this tool while HA is using the radio (since of the serial communication with the radio being exclusive) was quite key and helpful for me.
Just saw the pull request from Hedda to update the readme.md with more beginner oriented instructions [Update README.md with beginner instructions for NVRAM backup and restore](https://github.com/zigpy/zigpy-znp/pull/66) Will now go through them to see whether I can get it to work that way, probably starting on Windows.

**Reply by puddly** [reply]
The pull request doesn't contain valid info so I wouldn't do anything based on it until it's fixed.  Just follow the normal installation instructions for any other Python package. You may have to run `py -m pip` instead of `pip` and `py` instead of `python` since you're using windows.

**Reply by balu79** [reply]
Ok, got it, will try tomorrow, thanks again

**Reply by balu79** [reply]
I think I installed it correctly and I think I got it to work with a slightly modified command in Win 10, basically leaving out the -p parameter (after checking the possible arguments / usage in Windows). A json backup file of 31kb size was created, will check it further tomorrow.
python -m zigpy_znp.tools.nvram_read COM3 -o C:\temp\backup.json

**Reply by balu79** [reply]
Wow, it worked like a charm! This is of great value to me, appreciate the creation of the zigpy-znp tools and the extra help! Can I buy you a coffee or any other way to show my appreciation?

**Reply by Hedda** [reply]
@balu79 Maybe you can test if https://github.com/walthowd/znp-firmware works and if not help troubleshoot that there(?).

Update: Sorry, that does not yet look to be compatible with C2530 and CC2531 based adapters.

**Reply by balu79** [reply]
Would first have to install and set up Docker at my end as well as make myself familiar with it (didn't use it yet), so would need a bit of time, but sure, I can try that

**Reply by puddly** [reply]
> Can I buy you a coffee or any other way to show my appreciation?

Thank you! Not at the moment but I appreciate the offer.

> Maybe you can test if https://github.com/walthowd/znp-firmware works and if not help troubleshoot that there(?).

From what I understand, Docker and Windows COM ports do not mix. I couldn't find a way to expose them within the container so while the container will start, you will have no way to communicate with the device. Directly installing the Python package in a virtual environment (or globally) seems to be the only way to do this within Windows and macOS at the moment.
